Cleanup incremental decommit - Part 2

ch1bo opened the pull request:
Second part of promised cleanup on #1344. This holds not as crucial changes as the other and more refactorings than fixes.

* This should resolve most of the comments I had about #1344 and this also does the "clearing" of $\textcolor{red}{\\red}$ areas in the specfication.

* Validate a client decommit request against the `localUTxO`.

  We can (and should) check against the local ledger state in both cases as it's our local
  node's view of what is valid and what it would snapshot next anyways.

* Make AckSn processing emit a PartySignedSnapshot

  This matches with the Σ̂[j] ← σj assignment from the pseudo-code.
  This does not change the semantics, as the SnapshotConfirmed state
  changed event includes the full multi-signature anyways.

* Only request snapshot on ReqDec if no snapshot in flight

* Added a missing $\setminus outputs(tx_\omega)$ to the spec.

* Resolved "Spec Gap" comments in `HeadLogic`

* Refactored `CloseTx`, `ContestTx` and `DecrementTx` to take `ConfirmedSnapshot`s and remove redundant `closeUTxOToDecommit` from `CloseTx`.

* Simplify/shorten/reword incremental decommit how-to

* Removed various unused things

* Enhance / properly use `TxTraceSpec` to detect problems in contest transaction creation by modeling versions in `ModelSnapshot`.

---

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Haddocks updated
* [ ] No new TODOs introduced or explained herafter

ch1bo committed:
Document and re-order arguments of 'decrement'

ch1bo committed:
Remove unused Semigroup and Monoid of UTxOHash

ch1bo committed:
Simplify decommit how-to

ch1bo committed:
Expand changelog entry

ch1bo committed:
Remove decommitRequester

ch1bo committed:
Update haddocks and drop a redundant test

ch1bo committed:
Check decommit request against local ledger

The explanation of why checking against the confirmed ledger on client
submission does not hold up. We can (and should) check against the local
ledger state in both cases as it's our local node's view of what is
valid and what it would snapshot next anyways.

ch1bo committed:
Only request snapshot on ReqDec if no snapshot in flight

ch1bo committed:
Resolve some spec gap comments in head logic

ch1bo committed:
Make AckSn processing emit a PartySignedSnapshot

This matches with the Σ̂[j] ← σj assignment from the pseudo-code.
This does not change the semantics, as the SnapshotConfirmed state
changed event includes the full multi-signature anyways.

ch1bo committed:
Align HeadLogic closer with spec

ch1bo committed:
Combine requires when processing a ReqSn

ch1bo committed:
Add DecommitRecorded and DecommitFinalized to genStateChanged

ch1bo committed:
Remove unused SnapshotOutcome data types (and module)

ch1bo committed:
Consolidate Wait outcomes for decommit request processing

ch1bo committed:
Not assert starting state in HeadLogicSpec

ch1bo committed:
Remove redundant language pragmas

ch1bo committed:
Refactor CloseTx/close/closeTx and add docs to openVersion

ch1bo committed:
Also refactore ContestTx/contest/contestTx to match close

ch1bo committed:
Also refactor DecrementTx/decrement/decrementTx to take a ConfirmedSnapshot

This provides a more uniform interface across the chain layer.

ch1bo committed:
Untangle some healthyXXX functions

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -698,28 +671,20 @@ contestTx scriptRegistry vk Snapshot{number, utxo, utxoToDecommit, version} sig
           , headId = headIdToCurrencySymbol headId
           , contesters = contester : closedContesters
-          , version = toInteger version
+          , version = toInteger openVersion -- TODO: This was 'version' before and that was wrong. Which test can catch it?
```
This would be an interesting test to write still.

ch1bo replied:
Resolved in https://github.com/cardano-scaling/hydra/pull/1513/commits/df2fcf2b7d0f5c08ec15dbfd4a2cd3a0711b76f0

ch1bo committed:
Review and unmark red on-chain spec areas

This includes a re-ordering of clauses in checkDecrement, which did not
change semantics (but script hashes).

ch1bo committed:
Clear gaps and red markup in offchain

ch1bo committed:
Add AssertionFailed logic error outcome and use it onDecrementTx

This, at least, results in log lines if this situation occurrs.

v0d1ch reviewed (approved):
I noticed at least the decrement figure is not updated but we will revisit the spec anyway so I can correct it in the future pr. :tada:

ch1bo committed:
Fix TxTraceSpec by modelling snapshot versions

By having versions in the ModelSnapshot, we can correctly test the
head validator with contest transactions that try to set wrong
versions (see removed TODO).

ffakenz reviewed (commented):
👏
